### PR TITLE
AppUsers functionality

### PR DESCRIPTION
**Pull request by achunBOX**

### CODE REVIEW

**Comment by boxcla** [comment]
Hi @achunBOX, thanks for the pull request. Before we can merge it, we need you to sign our Contributor License Agreement. You can do so electronically here: http://opensource.box.com/cla

Once you have signed, just add a comment to this pull request saying, "CLA signed". Thanks!

File: BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.m [hunk 1]
```diff
@@ -77,6 +80,9 @@ + (BOXContentClient *)clientForUser:(BOXUserMini *)user
     @synchronized(synchronizer)
     {
         BOXContentClient *client = [[[self class] SDKClients] objectForKey:user.modelID];
+        
+        // Note: Places where uses can log into the same account as an OAuth and AppUser which causes a conflict in SDKClients
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.m]: Clean up this comments

File: BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.h [hunk 2]
```diff
@@ -55,6 +56,11 @@
 @property (nonatomic, readonly, strong) BOXUserMini *user;
 
 /**
+ * The delegate for the BOXContentClient instance (Internally BOXAPIQueueManager)
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.h]: Improve the comment here. What is this delegate used for? what does it do? (briefly)

File: BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.h [hunk 3]
```diff
@@ -6,6 +6,7 @@
 //
 
 #import "BOXOAuth2Session.h"
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.h]: Do we need to import these two here, or is it sufficient to import 'BOXAbstractSession'? Or even just forward declare it as a class

File: BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.m [hunk 4]
```diff
@@ -166,44 +174,53 @@ - (instancetype)initWithBOXUser:(BOXUserMini *)user
...
     {
-        [_OAuth2Session restoreCredentialsFromKeychainForUserWithID:user.modelID];
+        BOXOAuth2Session *session = [[BOXOAuth2Session alloc]init];
+        [session restoreCredentialsFromKeychainForUserWithID:user.modelID];
+        
+        if (session.refreshToken == nil) {
+            self.session = [[BOXAppUserSession alloc]initWithAPIBaseURL:self.APIBaseURL queueManager:self.queueManager];
+        }
+        
+        [self.session restoreCredentialsFromKeychainForUserWithID:user.modelID];
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Clients/BOXContentClient.m:178]: clean up

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIAppAuthOperation.h [hunk 5]
```diff
@@ -0,0 +1,74 @@
+//
+//  BOXAPIAppAuthToJSONOperation.h
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Operations/BOXAPIAppAuthOperation.h]: BOXAPIAppAuthOperation

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIAppUsersAuthOperation.h [hunk 6]
```diff
@@ -0,0 +1,74 @@
+//
+//  BOXAPIAppAuthToJSONOperation.h
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Operations/BOXAPIAppUsersAuthOperation.h]: BOXAPIAppUsersAuthOperation

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIAppAuthOperation.h (hunk 5 above)
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Operations/BOXAPIAppAuthOperation.h]: Remove this file, it isn't in the project and is redundant with BOXAPIAppUsersAuthOperation

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIAppUsersAuthOperation.h [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+/**
+ * Call either a success or failure callback depending on whether or not an error occured during the request.
+ */
+- (void)performCompletionCallback;
+
+/**
+ * Does nothing for BOXAPIAppAuthOperation and should never be called.
+ */
+- (void)processResponseData:(NSData *)data;
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Operations/BOXAPIAppUsersAuthOperation.h]: Do we need to declare this in the header?

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIAppAuthOperation.m [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        if (self.success) {
+            self.success(self.accessToken, self.accessTokenExpiration);
+        }
+    } else {
+        if (self.failure) {
+            self.failure(self.error);
+        }
+    }
+    
+    [[NSNotificationCenter defaultCenter]postNotificationName:BOXAuthOperationDidCompleteNotification object:self];
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Operations/BOXAPIAppAuthOperation.m]: Do we need this in addition to the login/refresh notifs?

File: BoxContentSDK/BoxContentSDK/Operations/BOXAPIAppUsersAuthOperation.m [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        if (self.success) {
+            self.success(self.accessToken, self.accessTokenExpiration);
+        }
+    } else {
+        if (self.failure) {
+            self.failure(self.error);
+        }
+    }
+    
+    [[NSNotificationCenter defaultCenter]postNotificationName:BOXAuthOperationDidCompleteNotification object:self];
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/Operations/BOXAPIAppUsersAuthOperation.m:79]: Do we need this in addition to the login/refresh notifs?

Comment (achunBOX) [inline BoxContentSDK/BoxContentSDK/Operations/BOXAPIAppUsersAuthOperation.m:79]: Yes, as it's being enqueued, the BOXAPIQueueManager will add itself as an observer to this notification and remove itself after it's authenticated.

**Comment by boxtcarpel** [comment]
+2 LGTM. Some comments

:busts_in_silhouette:

**Pull request merged**